specification/draft-2/README.md

```diff
@@ -0,0 +1,687 @@
+Identifiers for inputs and outputs must be unique across the document.
+
+The `binding` section of the output object contains a single property `glob`
+with a glob pattern describing how to recognize file generated by the tool.
+
+Additionally, `baseCmd` and `stdout` properties describe the executable name
+(optionally with additional hard-coded arguments like `python -m mymodule`)
+and standard output redirection (into a file called "output.txt").
+
+Success or failure of a job is determined by process exit code.
+By default, 0 is success and everything else is failure.
+However, some tools use exit codes to signal some information.
+The grep tool will exit with code 1 if no lines matched the pattern.
+For this purpose, we use the `successCodes` property to enumerate valid exit codes.
+
+Values for the `type` property can be any valid Apache Avro schema (in JSON form)
+with the `File` named type pre-defined. The `File` type declares that the input
+value should be passed in form of a file reference. At runtime, the file object
+itself will have at least the `path` and `size` properties, describing its
+local path and size in bytes.
+
+The File input and output of this tool are marked with the `streamable` flag to
+indicate that the tool will not seek over these files. CWL implementations might
+make use of this to pipe data from one tool to another.
+
+### Running tools
+
+A CWL implementation can run the above grep tool using a `Job` object like:
+
+```yaml
+inputs:
+  in_file:
+    __class: File
+    path: /local/path/to/file.txt
+    size: 42
+  pattern: find ?me
+allocatedResources:
+  cpu: 1
+  mem: 1024
+```
+
+The `inputs` property is a map that matches the input definitions of the tool.
+Files are encoded as objects with `__class` property set to `File` and with
+at least `path` and `size` properties.
+
+Additionally, `allocatedResources` section describes that the tool has been
+allocated a single CPU core and one GB of RAM.
+
+Running the grep tool with this job will create the following command line:
+
+```sh
+grep "find ?me" /local/path/to/file.txt > output.txt
+```
+
+After running the tool (in a directory specifically created for the job),
+resulting output object is created that looks like this:
+
+```yaml
+out_file:
+  __class: File
+  path: /path/to/job_directory/output.txt
+  size: 13
+```
+
+### Optional and named arguments
+
+Let's add some optional inputs to the grep tool: `--ignore-case` and `--context`
+
+```yaml
+# ...
+inputs:
+  # ...
+  - id: ignore_case
+    type: ['null', boolean]
+    binding:
+      prefix: --ignore-case
+  - id: context
+    type: ['null', integer]
+    binding:
+      prefix: --context
+# ...
+```
+
+Above, type of these two inputs is declared as union of "null" (note this must be a string)
+and their actual type to signify that these inputs are optional.
+
+Bindings include the `prefix` property that specifies these should be passed to
```

Comment:
No preference between prefix+separator and prefix+join. Slightly prefer these two to just prefix (with trailing space meaning join=false) since trailing spaces don't play well with visual editors, but this can be worked around.

On the other hand, if we use Scott's notation, we remove a whole property so that's good :)